geo/src/algorithm/stitch.rs

```diff
@@ -0,0 +1,590 @@
+use std::collections::HashMap;
+
+use geo_types::{Coord, Line, LineString, MultiPolygon, Polygon, Triangle};
+
+use crate::winding_order::WindingOrder;
+use crate::{Contains, GeoFloat};
+
+// ========= Error Type ============
+
+#[derive(Debug)]
+pub enum LineStitchingError {
+    IncompleteRing(&'static str),
+    InvalidGeometry(&'static str),
+    MissingParent,
+    NoExtremum,
+}
+
+impl std::fmt::Display for LineStitchingError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+impl std::error::Error for LineStitchingError {}
+
+pub(crate) type TriangleStitchingResult<T> = Result<T, LineStitchingError>;
+
+// ========= Main Algo ============
+
+/// Trait to stitch together split up triangles.
+pub trait StitchTriangles<T: GeoFloat> {
+    /// This stitching only happens along identical edges which are located in two separate
+    /// geometries. Please read about the required pre conditions of the inputs!
+    ///
+    /// ```text
+    /// ┌─────x        ┌─────┐
+    /// │    /│        │     │
+    /// │   / │        │     │
+    /// │  /  │  ───►  │     │
+    /// │ /   │        │     │
+    /// │/    │        │     │
+    /// x─────┘        └─────┘
+    /// ```
+    ///
+    /// # Pre Conditions
+    ///
+    /// - All the given triangles must be unique. If you're not sure about that, deduplicate before
+    ///   calling the algorithm!
+    /// - Input triangles should be valid polygons. For a definition of validity
+    ///   c.f. <https://www.postgis.net/workshops/postgis-intro/validity.html>
```

Comment:
That's kind of a chicken-and-egg problem imo. In the end, this is what `SpadeBoolops::union` is supposed to do.

Maybe it would make sense to not expose this trait to users at all, keep it private and only use it for SpadeBoolops? It's a pity but I also share your concerns. The preconditions could disappoint uninformed users and in a worst case scenario lead to hidden or hard to find bugs.

Reply:
Maybe... I'm not sure. I was truly just asking though, because I'm trying to understand what exactly I should expect the inputs and outputs to these methods to be. 

It sounds like it's true then, that a precondition of this trait is that the input triangles must not overlap. If so we should document that clearly.

A name like stitch_triangulation might better suggest the limitations of the method.

Reply:
Very much thinking out loud here but we do have a way of determining that two geometries don't overlap. There's a perf implication, but if we _don't_ want to make it internal only we could have something like a checked variant?

Reply:
Speaking personally, almost all of my own interest in this PR is the union implementation that depends on it. My own tack would be to make whatever simplifying assumptions you can to make this code work for the spade bool ops.

That said, I'm nobodies boss, so if you think there's more important value in iterating on making this triangle stitching trait stand alone, I'm willing to spend more time helping to review it.

In any case though, I need to understand what the code is *intended* to do before I can review it properly.

(also, thanks for your patience and good demeanor in the review process so far @RobWalt, I know it's kind of slow, but it's novel code that I'm still working towards understanding.)

Reply:
I was just trying to think a bit about the `checked` idea. Let me say that I do like it! On the other hand this would probably spawn a lot of other questions regarding the extra API, performance, etc. To get this PR over the finish line I'd like to go with the internal trait route. That is:

- make the trait internal to the `geo` crate (for now)
- document:
  - the assumptions regarding overlap a bit more for the after world
  - that the trait is currently only here to be used in the `SpadeBoolops` but that it can be used in other scenarios if it seems useful

We can always change it to be more public in the future if there is a real use case for it.

Let me know what you think!

---

@michaelkirk no worries. I'm thanking you for sticking with me, repeatingly reviewing and helping me to improve this so much already <3

Reply:
That sounds good to me!

Reply:
This should be resolved now